crates/types/src/vote.rs

```diff
@@ -255,7 +319,448 @@ pub trait Accumulator<T, U>: Sized {
     fn append(self, val: T) -> Either<Self, U>;
 }
 
+/// Accumulator trait used to accumulate votes into an `AssembledSignature`
+pub trait Accumulator2<
+    TYPES: NodeType,
+    COMMITTABLE: Committable + Serialize + Clone,
+    VOTE: VoteType<TYPES, COMMITTABLE>,
+>: Sized
+{
+    /// Append 1 vote to the accumulator.  If the threshold is not reached, return
+    /// the accumulator, else return the `AssembledSignature`
+    /// Only called from inside `accumulate_internal`
+    fn append(
+        self,
+        vote: VOTE,
+        vote_node_id: usize,
+        stake_table_entries: Vec<<TYPES::SignatureKey as SignatureKey>::StakeTableEntry>,
+    ) -> Either<Self, AssembledSignature<TYPES>>;
+}
+
+/// Accumulates DA votes
+pub struct DAVoteAccumulator<
+    TYPES: NodeType,
+    COMMITTABLE: Committable + Serialize + Clone,
+    VOTE: VoteType<TYPES, COMMITTABLE>,
+> {
+    /// Map of all da signatures accumlated so far
+    pub da_vote_outcomes: VoteMap<COMMITTABLE, TYPES::VoteTokenType>,
+    /// A quorum's worth of stake, generally 2f + 1
+    pub success_threshold: NonZeroU64,
+    /// A list of valid signatures for certificate aggregation
+    pub sig_lists: Vec<<BLSOverBN254CurveSignatureScheme as SignatureScheme>::Signature>,
+    /// A bitvec to indicate which node is active and send out a valid signature for certificate aggregation, this automatically do uniqueness check
+    pub signers: BitVec,
+    /// Phantom data to specify the vote this accumulator is for
+    pub phantom: PhantomData<VOTE>,
+}
+
+impl<
+        TYPES: NodeType,
+        COMMITTABLE: Committable + Serialize + Clone,
+        VOTE: VoteType<TYPES, COMMITTABLE>,
+    > Accumulator2<TYPES, COMMITTABLE, VOTE> for DAVoteAccumulator<TYPES, COMMITTABLE, VOTE>
+{
+    fn append(
+        mut self,
+        vote: VOTE,
+        vote_node_id: usize,
+        stake_table_entries: Vec<<TYPES::SignatureKey as SignatureKey>::StakeTableEntry>,
+    ) -> Either<Self, AssembledSignature<TYPES>> {
+        let VoteData::DA(vote_commitment) = vote.get_data() else {
+            return Either::Left(self);
+        };
+
+        let encoded_key = vote.get_key().to_bytes();
+
+        // Deserialize the signature so that it can be assembeld into a QC
+        // TODO ED Update this once we've gotten rid of EncodedSignature
```

Comment:
BTW, why do we have `EncodedSignature` originally? Will relaying the serialized format help speed up?